src/pubsub.js

```diff
@@ -0,0 +1,569 @@
+                (cb) => ipfs1.pubsub.ls(cb),
+                (cb) => {
+                  ipfs1.pubsub.unsubscribe(topic, handler2)
+                  cb()
+                },
+                (cb) => ipfs1.pubsub.ls(cb)
+              ], (err, res) => {
+                expect(err).to.not.exist
+
+                // Still subscribed as there is one listener left
+                expect(res[0]).to.be.eql([topic])
+                // Now all listeners are gone no subscription anymore
+                expect(res[2]).to.be.eql([])
+                check()
+              })
+            }
+
+            const handler2 = (msg) => {
+              expect(msg.data.toString()).to.be.eql('hello')
+              check()
+            }
+
+            parallel([
+              (cb) => ipfs1.pubsub.subscribe(topic, handler1, cb),
+              (cb) => ipfs1.pubsub.subscribe(topic, handler2, cb)
+            ], (err) => {
+              expect(err).to.not.exist
+              ipfs1.pubsub.publish(topic, 'hello', check)
+            })
+          })
+
+          it('discover options', (done) => {
+            const check = makeCheck(2, done)
+
+            const handler = (msg) => {
+              expect(msg.data.toString()).to.be.eql('hi')
+              ipfs1.pubsub.unsubscribe(topic, handler)
+              check()
+            }
+
+            ipfs1.pubsub.subscribe(topic, {
+              discover: true
+            }, handler, (err) => {
+              expect(err).to.not.exist
+              ipfs1.pubsub.publish(topic, 'hi', check)
+            })
+          })
+        })
+      })
+
+      describe('multiple nodes connected', () => {
+        before((done) => {
+          parallel([
+            (cb) => ipfs1.swarm.connect(ipfs2.peerId.addresses[0], cb),
+            (cb) => ipfs2.swarm.connect(ipfs3.peerId.addresses[0], cb),
+            (cb) => ipfs1.swarm.connect(ipfs3.peerId.addresses[0], cb)
+          ], done)
+        })
+
+        describe('.peers', () => {
+          it('returns an error when not subscribed to a topic', (done) => {
+            ipfs1.pubsub.peers(topic, (err, peers) => {
+              expect(
+                err.message
+              ).to.match(
+                new RegExp(`Not subscribed to '${topic}'`)
+              )
+
+              done()
+            })
+          })
+
+          it.skip("doesn't return extra peers", (done) => {
+            // Currently go-ipfs returns peers that have not been
+            // subscribed to the topic. Enable when go-ipfs has been fixed
+            const sub1 = (msg) => {}
+            const sub2 = (msg) => {}
+
+            const topicOther = topic + 'different topic'
+
+            series([
+              (cb) => ipfs1.pubsub.subscribe(topic, sub1, cb),
+              (cb) => ipfs2.pubsub.subscribe(topicOther, sub2, cb)
+            ], (err) => {
+              expect(err).to.not.exist
+              setTimeout(() => {
+                ipfs1.pubsub.peers(topic, (err, peers) => {
```

Comment:
It would be awesome to have a way to fetch peer info from the http api, namely known multiaddrs, public key, supported transports, etc //cc @whyrusleeping